File: 662-maximum-width-of-binary-tree/662-maximum-width-of-binary-tree.scala
import scala.collection.mutable._

          // 1
        // 2   3
      // 4  5 6  7 <-  6 - 3 + 1 = 4

object Solution {
    def widthOfBinaryTree(root: TreeNode): Int = {
        val queue: Queue[(TreeNode, Long)] = Queue()
        var res = 0l

        queue.enqueue((root, 1))

        while(queue.nonEmpty) {
            val size = queue.size

            var min = Long.MaxValue
            var max = Long.MinValue

            for(i <- 1 to size) {
                val (node, offset) = queue.dequeue
                if (node.left != null) queue.enqueue((node.left, offset * 2))
                if (node.right != null) queue.enqueue((node.right, offset * 2 + 1))
                min = offset min min
                max = offset max max
            }

            res = res max (max - min + 1)
        }

        res.asInstanceOf[Int]
    }
}

File: 662-maximum-width-of-binary-tree/662-maximum-width-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int widthOfBinaryTree(TreeNode root) {
        long res = 0;
        
        Queue<Node> queue = new LinkedList<>();
        
        if (root != null) queue.offer(new Node(root, 0));
        
        while(!queue.isEmpty()) {
            int sz = queue.size();
            
            long min = Long.MAX_VALUE;
            long max = Long.MIN_VALUE;
            
            for(int i = 0; i < sz; i++) {
                Node p = queue.poll();
                
                min = Math.min(p.offset, min);
                max = Math.max(p.offset, max);
                
                if (p.node.left != null) queue.offer(new Node(p.node.left, p.offset * 2));
                if (p.node.right != null) queue.offer(new Node(p.node.right, p.offset * 2 + 1));
            }
            
            res = Math.max(res, max - min + 1);
        }
        
        return (int)res;
    }
    
    class Node {
        public TreeNode node;
        public long offset;
        
        public Node(TreeNode node, long offset) {
            this.node = node;
            this.offset = offset;
        }
    }
}